pkg/api/dirs.go

```diff
@@ -224,98 +216,116 @@ func storeDir(ctx context.Context, encrypt bool, reader io.ReadCloser, log loggi
+		fileSize := fileHeader.FileInfo().Size()
+		filePath := filepath.Clean(fileHeader.Name)
 
-	// now join both references (fr, mr) to create an entry and store it
-	e := entry.New(manifestBytesReference, mr)
-	fileEntryBytes, err := e.MarshalBinary()
-	if err != nil {
-		return swarm.ZeroAddress, fmt.Errorf("entry marshal: %w", err)
-	}
+		if filePath == "." {
+			t.logger.Warning("skipping file upload empty path")
+			continue
+		}
+		if runtime.GOOS == "windows" {
+			// always use Unix path separator
+			filePath = filepath.ToSlash(filePath)
+		}
+		// only store regular files
+		if !fileHeader.FileInfo().Mode().IsRegular() {
+			t.logger.Warningf("skipping file upload for %s as it is not a regular file", filePath)
+			continue
+		}
 
-	manifestFileReference, err := p(ctx, bytes.NewReader(fileEntryBytes), int64(len(fileEntryBytes)))
-	if err != nil {
-		return swarm.ZeroAddress, fmt.Errorf("split entry: %w", err)
+		return &FileInfo{
+			Path:        filePath,
+			Name:        fileName,
+			ContentType: contentType,
+			Size:        fileSize,
+			Reader:      t.r,
+		}, nil
 	}
+}
 
-	return manifestFileReference, nil
+type multipartReader struct {
+	r      *multipart.Reader
+	logger logging.Logger
 }
 
-// storeFile uploads the given file and returns its reference
-// this function was extracted from `fileUploadHandler` and should eventually replace its current code
-func storeFile(ctx context.Context, fileInfo *fileUploadInfo, p pipelineFunc, encrypt bool, tag *tags.Tag, tagCreated bool) (swarm.Address, error) {
-	// first store the file and get its reference
-	fr, err := p(ctx, fileInfo.reader, fileInfo.size)
+func (m *multipartReader) Next() (*FileInfo, error) {
+	part, err := m.r.NextPart()
 	if err != nil {
-		return swarm.ZeroAddress, fmt.Errorf("split file: %w", err)
+		return nil, err
 	}
 
-	// if filename is still empty, use the file hash as the filename
-	if fileInfo.name == "" {
-		fileInfo.name = fr.String()
+	fileName := part.FileName()
+	if fileName == "" {
+		fileName = part.FormName()
 	}
-
-	// then store the metadata and get its reference
-	m := entry.NewMetadata(fileInfo.name)
-	m.MimeType = fileInfo.contentType
-	metadataBytes, err := json.Marshal(m)
-	if err != nil {
-		return swarm.ZeroAddress, fmt.Errorf("metadata marshal: %w", err)
+	if fileName == "" {
+		m.logger.Error("filename missing in multipart request")
+		return nil, errors.New("filename missing")
 	}
 
-	if !tagCreated {
-		// here we have additional chunks:
-		// - for metadata (1 or more) -> we use estimation function
-		// - for collection entry (1)
-		estimatedTotalChunks := calculateNumberOfChunks(int64(len(metadataBytes)), encrypt)
-		err = tag.IncN(tags.TotalChunks, estimatedTotalChunks+1)
-		if err != nil {
-			return swarm.ZeroAddress, fmt.Errorf("increment tag: %w", err)
-		}
+	contentType := part.Header.Get(contentTypeHeader)
+	if contentType == "" {
+		m.logger.Error("content-type missing in multipart request")
```

Comment:
no need to log and error and return it, the calling context should decide whether to log the error or propagate it further up the call stack